### PR TITLE
 Rank projects against each other

### PR DESCRIPTION
**Pull request by priyansh1913**
## Description
This PR adds a project ranking method to Project Report, enabling objective comparison using metrics such as lines of code, date range, and individual contribution. 
**Closes:** #157 

---

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation added/updated
- [X] Test added/updated
- [ ] Refactoring
- [ ] Performance improvement

---

## Testing

- [X] Test ranking calculation for projects with different lines, dates, and contributions
- [X] Test ranking when some statistics are missing or zero
- [X] Test ranking for projects with extreme values (max lines, max date range, max contribution)
- [X] Test ranking for projects with mixed individual and group contributions
- [X] Test ranking for projects with empty or no file reports
- [X] Test ranking for projects with undated or untitled reports
- [X] Test ranking for projects with similar statistics (should yield similar weights)
- [X] Test edge cases: negative values, missing fields, and empty database

---

## Checklist

- [X] GenAI was used in generating the code and I have performed a self-review of my own code
- [X] I have commented my code where needed
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## Screenshots

> If applicable, add screenshots to help explain your changes

<details>
<summary>Click to expand screenshots</summary>

</details>

### CODE REVIEW

File: src/classes/report.py [hunk 1]
```diff
@@ -120,6 +120,77 @@ def __init__(self,
...
+                    FileStatCollection.RATIO_OF_INDIVIDUAL_CONTRIBUTION.value) or 0.0
+                for report in self.file_reports
+            ]
+            if contribs:
+                contrib = sum(contribs) / len(contribs)
+        norm_contrib = min(max(float(contrib), 0.0) / 100.0,
+                           1.0) if contrib is not None else 0.0
+
+        # Linear combination with equal weights
+        weight = norm_lines + norm_date + norm_contrib
```
Comment (thndlovu) [inline src/classes/report.py]: The docstring implies the whole weight is in [0, 1], since it says all factors are normalized to [0, 1] and are given equal weight. But the actual implementation sums the three normalized components, so the returned value is in [0, 3].

Could you clarify this in the docstring, or divide the final sum by 3 so the method itself returns a [0, 1] score?

**Review by thndlovu — CHANGES_REQUESTED** [review]
Thanks for the PR, the implementation and tests look solid overall. I left one requested change on the docstring to clarify the expected range of the project weight (whether it’s [0, 1] or [0, 3]). Once that’s addressed, I’m happy to approve.

**Review by sjsikora — APPROVED** [review]
Thanks Priyansh for your work on this! I had some comments to make sure the code is as clean as possible in this func.

**Review by sjsikora — CHANGES_REQUESTED** [review]
See above comments, meant to request changes

File: src/classes/report.py [hunk 2]
```diff
@@ -120,6 +120,77 @@ def __init__(self,
         # Initialize the base class with the project statistics
         super().__init__(self.project_statistics)
 
+    def get_project_weight(self) -> float:
+        """
+        Ranks the project using a linear combination of lines of code, date range, and individual contribution level.
+        Equal weightage is given to each factor. All factors are normalized to [0, 1] scale.
+        """
+        total_lines = 0.0
+        if hasattr(self, 'file_reports') and self.file_reports:
```
[the remark this answers is not in the document]
Comment (priyansh1913) [reply src/classes/report.py]: Ya I have removed the check in the latest commit, as you are right the report will always have a file_report

**Review by sjsikora — APPROVED** [review]
Thanks for all those changes!

**Review by thndlovu — APPROVED** [review]
Thanks for addressing the docstring clarification. Great implementation, LGTM!
